memory_util.h

```diff
@@ -0,0 +1,298 @@
+    data_ = (size != 0 ? new uchar[size] : NULL);
+    size_ = capacity_ = size;
+    own_ = true;
+}
+
+inline BinaryBuffer::BinaryBuffer(size_type size, size_type capacity) {
+    if (size > capacity) {
+        size = capacity;
+    }
+
+    data_ = (capacity != 0 ? new uchar[capacity] : NULL);
+    size_ = size;
+    capacity_ = capacity;
+    own_ = true;
+}
+
+inline BinaryBuffer::BinaryBuffer(const void* data, size_type size) {
+    if (size != 0) {
+        data_ = new uchar[size];
+        std::memcpy(data_, data, size);
+    } else {
+        data_ = NULL;
+    }
+    size_ = capacity_ = size;
+    own_ = true;
+}
+
+inline BinaryBuffer::BinaryBuffer(const BinaryBuffer& other) {
+    if (other.capacity_) {
+        data_ = new uchar[other.capacity_];
+        if (other.size_) {
+            std::memcpy(data_, other.data_, other.size_);
+        }
+    } else {
+        data_ = NULL;
+    }
+    size_ = other.size_;
+    capacity_ = other.capacity_;
+    own_ = true;
+}
+
+inline BinaryBuffer& BinaryBuffer::operator=(const BinaryBuffer& other) {
+    if (&other != this) {
+        if (other.size_ > capacity_) {
+            delete_if_own();
+            data_ = new uchar[other.capacity_];
+            capacity_ = other.capacity_;
+        }
+        std::memcpy(data_, other.data_, other.size_);
+        size_ = other.size_;
+    }
+    return *this;
+}
+
+inline void BinaryBuffer::assign(const void* data, size_type size) {
+    if (size > capacity_) {
+        delete_if_own();
+        data_ = new uchar[size];
+        capacity_ = size;
+        own_ = true;
+    }
+    std::memcpy(data_, data, size);
+    size_ = size;
+}
+
+inline void BinaryBuffer::take_over(void* data, size_type size) {
+    delete_if_own();
+    data_ = static_cast<uchar*>(data);
+    size_ = capacity_ = size;
+    own_ = true;
+}
+
+inline void BinaryBuffer::append(const BinaryBuffer& other) {
+    append(other.data_, other.size_);
+}
+
+template <typename T>
+inline void BinaryBuffer::append(const T t) {
+   append(&t, 1);
+}
+
+inline void BinaryBuffer::append(const std::string& str) {
+    append((std::size_t)str.length());
+    append(str.c_str(), str.length()+1);
+}
+
+template <typename T>
+inline void BinaryBuffer::append(const std::vector<T>& vec) {
+    append((std::size_t)vec.size());
+    append(vec.data(), vec.size());
+}
+
+template <typename T>
+inline void BinaryBuffer::append(const T* data, int num_elements) {
+    if (num_elements == 0) {
+        return;
+    }
+
+    size_type total_size = sizeof(T) * num_elements;
+    reserve(total_size + size_);
+    std::memcpy(data_ + size_, (uchar*)data, total_size);
+    size_ = size_ + total_size;
+}
+
+inline void BinaryBuffer::start_reading() {
+    read_ptr_ = data_;
+}
+
+template <typename T>
+inline void BinaryBuffer::read(T& t) {
+    read(&t, 1);
+}
+
+inline void BinaryBuffer::read(std::string& str) {
+    std::size_t len;
+    read(len);
+    char* tmp = new char[len+1];
```

Comment:
You can just use stack-based mem allocation.
"char tmp[len+1];"
Then you don't have to delete [] tmp.